perceval/backends/mediawiki.py

```diff
@@ -0,0 +1,517 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#     Alvaro del Castillo <acs@bitergia.com>
+#
+
+import datetime
+import json
+import logging
+import os.path
+
+import dateutil
+import requests
+
+from ..backend import Backend, BackendCommand, metadata
+from ..cache import Cache
+from ..errors import BackendError, CacheError
+from ..utils import (DEFAULT_DATETIME,
+                     datetime_to_utc,
+                     str_to_datetime,
+                     datetime_to_utc,
+                     urljoin)
+
+
+logger = logging.getLogger(__name__)
+
+MAX_RECENT_DAYS = 30  # max number of days included in MediaWiki recent changes
+
+
+class MediaWiki(Backend):
+    """MediaWiki backend for Perceval.
+
+    This class retrieves the wiki pages and edits from a
+    MediaWiki site. To initialize this class the
+    URL must be provided.
+
+    :param url: MediaWiki url
+    :param cache: cache object to store raw data
+    :param origin: identifier of the repository; when `None` or an
+        empty string are given, it will be set to `url` value
+    """
+    version = '0.1.0'
+
+    def __init__(self, url, cache=None, origin=None):
+        origin = origin if origin else url
+
+        super().__init__(origin, cache=cache)
+        self.url = url
+        self.client = MediaWikiClient(url)
+
+    @metadata
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the pages from the backend url.
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        self._purge_cache_queue()
+
+        if from_date == DEFAULT_DATETIME:
+            from_date = None
+        else:
+            from_date = datetime_to_utc(from_date).replace(tzinfo=None)
+
+        mediawiki_version = self.client.get_version()
+        logger.info("MediaWiki version: %s", mediawiki_version)
+
+        if (mediawiki_version[0] == 1 and mediawiki_version[1] >= 27) or \
+            mediawiki_version[0] > 1:
+            fetcher = self.__fetch_1_27(from_date)
+        else:
+            fetcher = self.__fetch_pre1_27(from_date)
+
+        for page_reviews in fetcher:
+            yield page_reviews
+
+
+    @metadata
+    def fetch_from_cache(self):
+        """Fetch the pages from the cache.
+
+        :returns: a generator of pages
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        pages_dicts = ['allrevisions','allpages','recentchanges']
+
+        # Namespaces -> pages -> revisions
+        next(cache_items) # First namespaces, we don't need them
```

Comment:
Good point. I thought the same but in order to understand how the pages were retrieved it could be useful (for debugging the cache for example). But we can remove it, yes.

So, do you prefer removing them? I think it is the best option.